parlai/scripts/self_chat.py

```diff
@@ -30,92 +28,97 @@ def setup_args(parser=None):
     WorldLogger.add_cmdline_args(parser)
     return parser
 
 
-def self_chat(opt, print_parser=None):
-    if print_parser is not None:
-        if print_parser is True and isinstance(opt, ParlaiParser):
-            print_parser = opt
-        elif print_parser is False:
-            print_parser = None
-    if isinstance(opt, ParlaiParser):
-        print('[ Deprecated Warning: self_chat should be passed opt not Parser ]')
-        opt = opt.parse_args()
+def _run_self_chat_episode(opt, world, world_logger):
+    bsz = opt.get('batchsize', 1)
+    num_turns = opt['selfchat_max_turns']
+
+    num_parleys = math.ceil(num_turns / bsz)
+    for _ in range(num_parleys):
+        world.parley()
+        world_logger.log(world)
+
+        if opt['display_examples']:
+            print(world.display())
+
+    if opt['display_examples']:
+        print('-- end of episode --')
+
+    world_logger.reset_world()  # flush this episode
+
 
+def self_chat(opt):
     random.seed(opt['seed'])
-    # Create models
+
+    # Create agents
     agent1 = create_agent(opt, requireModelExists=True)
     agent2 = agent1.clone()
-    if hasattr(agent2, 'id'):
-        agent2.id = agent2.id + "2"
-
-    # Check for `selfchat` in the task name
-    if 'selfchat' not in opt['task']:
-        warn_once(
-            'You are using self chat with task {}. '.format(opt['task'])
-            + 'If your task has an existing self chat world, then run with '
-            '-t {}:selfchat'.format(opt['task'])
-        )
-
-    world = create_task(opt, [agent1, agent2])
-
-    if print_parser:
-        # Show arguments after loading model
-        print_parser.opt = agent1.opt
-        print_parser.print_args()
-
-    # set up logging
-    log_every_n_secs = opt.get('log_every_n_secs', -1)
-    if log_every_n_secs <= 0:
-        log_every_n_secs = float('inf')
-    log_time = TimeLogger()
-    logger = WorldLogger(opt)
 
-    # Run some self chats.
-    max_cnt = opt['num_examples']
-    cnt = 0
-    while cnt < max_cnt:
-        cnt += opt.get('batchsize', 1)
-        world.parley()
-        logger.log(world)
+    # Set IDs
+    model_id = agent1.id
+    agent1.id = model_id + "_1"
+    agent2.id = model_id + "_2"
 
-        if opt.get('display_examples'):
-            print(world.display())
-        if log_time.time() > log_every_n_secs:
-            text = log_time.log(cnt, max_cnt)
-            print(text)
+    world = create_task(opt, user_agents=[agent1, agent2])
 
-    if opt.get('display_examples'):
-        print('-- end of episode --')
+    # Set up world logging
+    logger = WorldLogger(opt)
+
+    # Run some self chats.
+    for _ in range(opt['num_self_chats']):
+        _run_self_chat_episode(opt, world, logger)
+
+    # Save chats
+    if opt['outfile'] is None:
+        outfile = '/tmp/{}_selfchat'.format(model_id)
+    else:
+        outfile = opt['outfile']
+
+    if opt['save_format'] == 'conversations' and hasattr(world, 'write'):
```

Comment:
so we only use `world.write` if the `save_format` is `conversations`? should this maybe be a separate option from `conversations`? or are we only supporting `world.write` with `conversations`, assuming that `world.write` will write in the conversation format?